File: builder_pattern.py
"""
focuses on constructing a simple immutable object, 
like a user profile, with several optional fields. 
This best illustrates why the pattern is needed—to handle many optional parameters without complex,
 hard-to-read constructors.

real-world applications include:
- Configuring complex objects like database connections or network requests.
- Building UI components with various optional settings.    

"""

# 1. Product (The complex object)
class User:
    def __init__(self, name, age=None, email=None, is_verified=False):
        self.name = name
        self.age = age
        self.email = email
        self.is_verified = is_verified

    def __str__(self):
        details = [f"Name: {self.name}"]
        if self.age is not None:
            details.append(f"Age: {self.age}")
        if self.email is not None:
            details.append(f"Email: {self.email}")
        details.append(f"Verified: {self.is_verified}")
        return ", ".join(details)

# 2. Concrete Builder
class UserBuilder:
    def __init__(self, name):
        # Mandatory field set in the Builder's constructor
        self._name = name
        # Optional fields initialized to None or default
        self._age = None
        self._email = None
        self._is_verified = False

    def set_age(self, age):
        self._age = age
        return self  # Return self for method chaining (Fluent API)

    def set_email(self, email):
        self._email = email
        return self

    def verify(self):
        self._is_verified = True
        return self

    def build(self):
        # The final step that creates and returns the immutable User object
        return User(
            name=self._name,
            age=self._age,
            email=self._email,
            is_verified=self._is_verified
        )

# Client Usage
# Building a fully configured user
user1 = UserBuilder("Alice").set_age(30).set_email("alice@example.com").verify().build()
print(f"User 1: {user1}")

# Building a minimally configured user
user2 = UserBuilder("Bob").set_age(25).build()
print(f"User 2: {user2}")

File: builder_pattern.go
package main

import "fmt"

// 1. Product Struct (The complex object to be built)
type Config struct {
	Host       string
	Port       int
	TimeoutSec int
	UseTLS     bool
}

// 2. Concrete Builder Struct
type ConfigBuilder struct {
	config Config
}

// Constructor for the Concrete Builder (takes mandatory fields)
func NewConfigBuilder(host string) *ConfigBuilder {
	return &ConfigBuilder{
		config: Config{
			Host:       host,
			Port:       8080, // Set default for optional Port
			TimeoutSec: 30,   // Set default for optional Timeout
			UseTLS:     false,
		},
	}
}

// Builder Step: Set Port (Returns the builder struct for chaining)
func (b *ConfigBuilder) SetPort(port int) *ConfigBuilder {
	b.config.Port = port
	return b
}

// Builder Step: Set Timeout
func (b *ConfigBuilder) SetTimeout(sec int) *ConfigBuilder {
	b.config.TimeoutSec = sec
	return b
}

// Builder Step: Enable TLS
func (b *ConfigBuilder) EnableTLS() *ConfigBuilder {
	b.config.UseTLS = true
	return b
}

// Build Method (Finalizes the product)
func (b *ConfigBuilder) Build() Config {
	// Optional: Validation logic can go here before returning
	if b.config.Host == "" {
		panic("Host is required!")
	}
	return b.config
}

func main() {
	// Building a custom configuration
	customConfig := NewConfigBuilder("api.service.com").
		SetPort(443).
		SetTimeout(60).
		EnableTLS().
		Build()

	fmt.Printf("Custom Config: %+v\n", customConfig)

	// Building a configuration using only defaults
	defaultConfig := NewConfigBuilder("localhost").Build()

	fmt.Printf("Default Config: %+v\n", defaultConfig)
}
